### ISSUE TITLE
"Show VLANs" on VLAN Membership tab do nothing.

### ISSUE THREAD

**Issue by fxthiry** [netdisco/netdisco#713]
Hello,

When i click to "Show VLANs"  to see all the VLAN's on a port, nothing happens.

<img width="234" alt="Screenshot 2020-04-14 at 13 40 32" src="https://user-images.githubusercontent.com/23431839/79221130-911ae400-7e55-11ea-997d-5dcb61592545.png">
<img width="234" alt="Screenshot 2020-04-14 at 13 40 25" src="https://user-images.githubusercontent.com/23431839/79221134-924c1100-7e55-11ea-904e-f9d47aa81097.png">

Is that normal ?

Thanks you.

**Comment by rc9000** [comment]
Hi! No that's not quite normal, I just checked with the current release and it works for me in all three modes - show, hide, too many:  

<img width="1142" alt="scr 2020-05-02 at 02 41 56" src="https://user-images.githubusercontent.com/1247031/80850944-f412c600-8c1e-11ea-80dc-30fde261f8de.png">

I also don't quite get how it could go wrong, since the vlans seem to be in the database given the number, and the list is just simple links that should be included in the AJAX response. If you look at the HTML source of the page, do you see the whole list of VLANs in there? E.g. in my browser, searching for Ethernet10/23 I get this table row (fragment `<tr>...</tr>`) with all 46 VLAN:

        <tr>
          <td class="nd_center-cell nd_devport-icon">
            <i class="icon-angle-up text-success icon-large"></i><br/>
                <small><i class="icon-group text-success"></i></small>
          </td>
          <td nowrap data-order="Ethernet10/23" data-filter="Ethernet10/23">
            <a class="nd_log-icon"
              href="/netdisco/report/portlog?q=192.168.227.89&f=Ethernet10%2F23">
              <i class="icon-file-text-alt"
                rel="tooltip" data-placement="top" data-offset="3"
                data-animation="" data-title="View Port Log"></i>
            </a>
            <a class="nd_this-port-only nd_port-only-first" href="/netdisco/device?tab=ports&c_pvid=on&c_stp=&n_inventory=on&n_archived=&age_num=3&c_ssid=&c_nodes=on&age_unit=months&c_lastchange=&c_name=on&c_admin=&c_mac=&c_duplex=&c_neighbors=&c_mtu=&n_vendor=&c_power=&c_up=&n_ip6=on&c_speed_admin=&n_detailed_inventory=&c_vmember=on&c_descr=&n_age=&n_netbios=on&n_ssid=on&c_port=on&c_error=&c_speed=&mac_format=IEEE&n_ip4=on&c_comment=&c_type=&q=malaga-sgc-x1002-vdc4.nms.admin.ch&f=Ethernet10%2F23&prefer=port">Ethernet10/23</a><br/>
              <a class="nd_this-port-only" href="/netdisco/device?tab=ports&c_pvid=on&c_stp=&n_inventory=on&n_archived=&age_num=3&c_ssid=&c_nodes=on&age_unit=months&c_lastchange=&c_name=on&c_admin=&c_mac=&c_duplex=&c_neighbors=&c_mtu=&n_vendor=&c_power=&c_up=&n_ip6=on&c_speed_admin=&n_detailed_inventory=&c_vmember=on&c_descr=&n_age=&n_netbios=on&n_ssid=on&c_port=on&c_error=&c_speed=&mac_format=IEEE&n_ip4=on&c_comment=&c_type=&q=malaga-sgc-x1002-vdc4.nms.admin.ch&f=port-channel318&prefer=port">port-channel318</a>
          </td>
          <td nowrap class="nd_editable-cell">
          <div class="nd_editable-cell-content">to malaga-sgc-x901-2 e1/24
          </div>
          </td>
          <td>
            <a class="nd_linkcell"
              href="/netdisco/search?tab=vlan&q=1">1</a>
          </td>
          <td> <div class="nd_vlan-total">(46)</div><span class="nd_linkcell nd_collapse-vlans">
                    <div class="nd_arrow-up-down-left icon-chevron-up icon-large"></div>Show VLANs</span>
                    <div class="nd_collapsing nd_collapse-pre-hidden"><a href="/netdisco/search?tab=vlan&q=1">1</a>, <a href="/netdisco/search?tab=vlan&q=767">767</a>, <a href="/netdisco/search?tab=vlan&q=865">865</a>, <a href="/netdisco/search?tab=vlan&q=866">866</a>, <a href="/netdisco/search?tab=vlan&q=867">867</a>, <a href="/netdisco/search?tab=vlan&q=868">868</a>, <a href="/netdisco/search?tab=vlan&q=869">869</a>, <a href="/netdisco/search?tab=vlan&q=1813">1813</a>, <a href="/netdisco/search?tab=vlan&q=1814">1814</a>, <a href="/netdisco/search?tab=vlan&q=1860">1860</a>, <a href="/netdisco/search?tab=vlan&q=1865">1865</a>, <a href="/netdisco/search?tab=vlan&q=1866">1866</a>, <a href="/netdisco/search?tab=vlan&q=1867">1867</a>, <a href="/netdisco/search?tab=vlan&q=1868">1868</a>, <a href="/netdisco/search?tab=vlan&q=1869">1869</a>, <a href="/netdisco/search?tab=vlan&q=1901">1901</a>, <a href="/netdisco/search?tab=vlan&q=1902">1902</a>, <a href="/netdisco/search?tab=vlan&q=1903">1903</a>, <a href="/netdisco/search?tab=vlan&q=1904">1904</a>, <a href="/netdisco/search?tab=vlan&q=1905">1905</a>, <a href="/netdisco/search?tab=vlan&q=1908">1908</a>, <a href="/netdisco/search?tab=vlan&q=1909">1909</a>, <a href="/netdisco/search?tab=vlan&q=1912">1912</a>, <a href="/netdisco/search?tab=vlan&q=1913">1913</a>, <a href="/netdisco/search?tab=vlan&q=1914">1914</a>, <a href="/netdisco/search?tab=vlan&q=1915">1915</a>, <a href="/netdisco/search?tab=vlan&q=2101">2101</a>, <a href="/netdisco/search?tab=vlan&q=2102">2102</a>, <a href="/netdisco/search?tab=vlan&q=2103">2103</a>, <a href="/netdisco/search?tab=vlan&q=2104">2104</a>, <a href="/netdisco/search?tab=vlan&q=2105">2105</a>, <a href="/netdisco/search?tab=vlan&q=2106">2106</a>, <a href="/netdisco/search?tab=vlan&q=2107">2107</a>, <a href="/netdisco/search?tab=vlan&q=2108">2108</a>, <a href="/netdisco/search?tab=vlan&q=2109">2109</a>, <a href="/netdisco/search?tab=vlan&q=2110">2110</a>, <a href="/netdisco/search?tab=vlan&q=2111">2111</a>, <a href="/netdisco/search?tab=vlan&q=2127">2127</a>, <a href="/netdisco/search?tab=vlan&q=2128">2128</a>, <a href="/netdisco/search?tab=vlan&q=2129">2129</a>, <a href="/netdisco/search?tab=vlan&q=2130">2130</a>, <a href="/netdisco/search?tab=vlan&q=2131">2131</a>, <a href="/netdisco/search?tab=vlan&q=2658">2658</a>, <a href="/netdisco/search?tab=vlan&q=2685">2685</a>, <a href="/netdisco/search?tab=vlan&q=3301">3301</a>, <a href="/netdisco/search?tab=vlan&q=3302">3302</a></div>
          </td>
          <td>
          </td>
        </tr>

Do you have that? Or if not, if you open the browser developer tools (e.g. F12 in Firefox), do you see any errors there in the Console tab?

Also if you're not on the last version, please upgrade first and make sure to restart the netdisco-web process.

**Comment by fxthiry** [comment]
Hi, thanks you for your response.

No, i have not <tr>...</tr> but i have the last version.

The errors / warnings :

> Some cookies are misusing the recommended “sameSite“ attribute 4

> Layout was forced before the page was fully loaded. If stylesheets are not yet loaded this may cause a flash of unstyled content. jquery-latest.min.js:6:26142

> This page uses the non standard property “zoom”. Consider using calc() in the relevant property values, or using “transform” along with “transform-origin: 0 0”. device

> Empty string passed to getElementById(). 3 jquery-latest.min.js:4:1384

> **XML Parsing Error: not well-formed**
> **Location: http://netdisco.xxx.com:5000/ajax/content/devic…ort_state=up&port_state=free&port_state=down&port_state=shut**
> **Line Number 26, Column 18: ports:26:18**

> Empty string passed to getElementById(). 6 jquery-latest.min.js:4:1384

> Source map error: Error: request failed with status 404
> Resource URL: http://netdisco.xxx.com:5000/javascripts/jquery-latest.min.js
> Source Map URL: jquery.min.map

> Empty string passed to getElementById(). 9 jquery-latest.min.js:4:1384
> Using //@ to indicate sourceMappingURL pragmas is deprecated. Use //# instead
> jquery-latest.min.js:2:21

> Source map error: Error: request failed with status 404
> Resource URL: http://netdisco.xxxx.com:5000/javascripts/jquery-latest.min.js
> Source Map URL: jquery.min.map
 
> Source map error: Error: request failed with status 404
> Resource URL: http://netdisco.xxxx.com:5000/javascripts/underscore.min.js
> Source Map URL: underscore-min.map

> Source map error: Error: request failed with status 404
> Resource URL: http://netdisco.xxxx.com:5000/javascripts/bootstrap2-toggle.min.js
> Source Map URL: bootstrap2-toggle.min.js.map

**Comment by rc9000** [comment]
Hmm interesting, can you attach what is sent back to you in the XML Parsing Error URL?

**Comment by fxthiry** [comment]
For security reason, i cannot do that, but its weird the line 26 is the only one duplicate (but different ip address) :

Line 26 : 

```GigabitEthernet1/24,XXX-XXX,1,"1,101,421,422,423,424,425,426,427,428,429,430,431,500,700,4000,4001,4002,4003",10.255.43.254,```

Line 27 :

```GigabitEthernet1/24,XXX-XXX,1,"1,101,421,422,423,424,425,426,427,428,429,430,431,500,700,4000,4001,4002,4003",10.255.40.254,```

**Comment by rc9000** [comment]
Ok I had a look in my installation and unfortunately we're not even looking at the problem yet. 

It's a bit tricky, the URL with the XML Error has a dynamic response. With just a GET in the browser it will send the CSV content you posted. When it's called from the GUI it will respond with the HTML variant of the port list. But the thing is, I'm also getting the error you posted on pages that work perfectly fine - there is just a configuration issue that the HTML variant declares XML as response type, and the error is just the developer tools trying to be helpful and parse that alleged XML.

Not quite sure what to do from here since I can't reproduce it, nobody else seems to have input  and you also seem to be the only person with this issue so far - dire times :)

 If you want to poke yourself around a bit, first make sure to upgrade to the latest version, then the code is here:

 * ./lib/App/Netdisco/Web/Plugin/Device/Ports.pm - since you say you don't even have the vlan list in the HTML, but you have the number of vlans, the bit around line 107 might somehow go wrong
 * share/views/ajax/device/ports.tt  is the corresponding template toolkit file

If you just put some debug output in there you can maybe find where it fails. The [Developing](https://github.com/netdisco/netdisco/wiki/Developing) section in the wiki has some good information on how to run Netdisco from the git directory, DBIC ORM essentials etc.

**Comment by ollyg** [comment]
One possible cause is that the loop creating the list of VLAN hyperlinks is creating invalid HTML in some way. It's this line, here:
https://github.com/netdisco/netdisco/blob/24a6c88a60bfe01b48056f21bd9f4adddde7971c/share/views/ajax/device/ports.tt#L221-L222

Removing these lines from `views/ajax/device/ports.tt` in `~nertdisco/perl5/...` would reveal the issue. If you get a set of commas (the text between each VLAN hyperlink) then we know the issue is in that code. If you still have the problem then we need to look elsewhere.

**Comment by izomorph** [comment]
I think problem in the line 223:
`[% SET output = output _ ', ' IF NOT loop.last %]`
Removing `IF NOT loop.last` helps (except additional comma at the end)

**Comment by ollyg** [comment]
Hi @izomorph thank you!
Could you try setting that line to be something like:
````
[% IF NOT loop.last %][% SET output = output _ ', ' %][% END %]
````

As I can't reproduce the error here, we need to go through this trial and error approach. Thanks for your patience.

**Comment by fxthiry** [comment]
Hello thank you all !

Its working.

![Screenshot 2020-05-18 at 11 35 25](https://user-images.githubusercontent.com/23431839/82197887-cbb3e880-98fb-11ea-9796-baead91a9941.png)

**Comment by izomorph** [comment]
Yes, its working now as expected. Thanks!

**Comment by ollyg** [comment]
Super, thank you @fxthiry and @izomorph !
I've released 2.045005 to CPAN with this fix.
[master abc8bb3e] #713 Template processing issue prevents VLAN list displaying
